File: src/player_context.c
#include <inttypes.h>
#include <string.h>
#include <assert.h>

#include "player_context.h"

void init_player_context (player_context_t *player)
{
   memset (player, 0, sizeof (player_context_t));
   player->player_number = 1;
   player->current_level = 0;
   player->score = 0;
   player->time = 1000;
   player->bonus = 900;
   player->lives = 5;
}

// player related setters and getters
uint8_t get_player_number (player_context_t *player)
{
   return (player->player_number);
}

void set_player_number (player_context_t *player, uint8_t num)
{
   player->player_number = num;
}

// level related setters and getters
uint8_t get_current_level (player_context_t *player)
{
   return (player->current_level);
}

void set_current_level (player_context_t *player, uint8_t num)
{
   player->current_level = num;
   set_current_level_gfx (player);
}

uint8_t *get_current_level_gfx (player_context_t *player)
{
   return (&player->gfx_current_level[0]);
}

void convert_num_to_gfx (uint32_t num, uint8_t *gfx,
                         uint32_t base, uint8_t level)
{
   uint32_t div = base;
   uint32_t helper = num;
   uint8_t i;

   for (i = 0; i < level; i++)
   {
      gfx[i] = helper / div;
      helper -= gfx[i] * div;
      div /= 10;
   }
}

void set_current_level_gfx (player_context_t *player)
{
   convert_num_to_gfx (player->current_level,
                       &player->gfx_current_level[0], 100, 3);
}

// score related setters and getters
uint32_t get_score (player_context_t *player)
{
   return (player->score);
}

void set_score (player_context_t *player, uint32_t num)
{
   player->score = num;
   set_score_gfx (player);
}

uint8_t *get_score_gfx (player_context_t *player)
{
   return (&player->gfx_score[0]);
}

void set_score_gfx (player_context_t *player)
{
   uint8_t prior = player->gfx_score[1];

   convert_num_to_gfx (player->score, &player->gfx_score[0], 100000, 6);

   // new life every 10000 points
   if (prior != player->gfx_score[1])
      // max is 255 lives
      if (get_lives (player) < 0xff)
         set_lives (player, get_lives (player) + 1);
}

// time related setters and getters
uint16_t get_time (player_context_t *player)
{
   return (player->time);
}

void set_time (player_context_t *player, uint16_t num)
{
   player->time = num;
   set_time_gfx (player);
}

uint8_t *get_time_gfx (player_context_t *player)
{
   return (&player->gfx_time[0]);
}

void set_time_gfx (player_context_t *player)
{
   convert_num_to_gfx (player->time, &player->gfx_time[0], 100, 3);
}

// bonus related setters and getters
uint16_t get_bonus (player_context_t *player)
{
   return (player->bonus);
}

void set_bonus (player_context_t *player, uint16_t num)
{
   player->bonus = num;
   set_bonus_gfx (player);
}

uint8_t *get_bonus_gfx (player_context_t *player)
{
   return (&player->gfx_bonus[0]);
}

void set_bonus_gfx (player_context_t *player)
{
   convert_num_to_gfx (player->bonus, &player->gfx_bonus[0], 1000, 4);
}

// lives related setters and getters
uint8_t get_lives (player_context_t *player)
{
   return (player->lives);
}

void set_lives (player_context_t *player, uint8_t num)
{
   player->lives = num;
}

uint8_t get_sandbox (player_context_t *player, uint8_t x, uint8_t y)
{
   return (player->sandbox[y][x]);
}

void set_sandbox (player_context_t *player, uint8_t x,
                  uint8_t y, uint8_t content)
{
   assert ((y < OFFSET_Y_MAX) && (x < OFFSET_X_MAX));

   player->sandbox[y][x] = content;
}

uint8_t get_n_eggs (player_context_t *player)
{
   return (player->n_of_eggs);
}

void set_n_eggs (player_context_t *player, uint8_t num)
{
   player->n_of_eggs = num;
}

File: py/player_context.py
class player_context_t:
   def __init__ (self):
      # player number 1 .. 4
      self.player_number = 1
      self.current_level = 0
      self.gfx_current_level = []
      # score 0 .. 999990 inclusive
      self.score = 0
      self.gfx_score = [0, 0, 0, 0, 0, 0]
      # time 0 .. 900 inclusive
      self.time = 1000
      self.gfx_time = []
      # bonus 0 .. 9000
      self.bonus = 900
      self.gfx_bonus = []
      # lives new life every 10000 score points max 255
      self.lives = 5
      # movement context
      self.sandbox = {}
      # collected eggs
      self.n_of_eggs = 0

   def get_player_number (self):
      return self.player_number
   def set_player_number (self, num):
      self.player_number = num

   def get_current_level (self):
      return self.current_level
   def set_current_level (self, num):
      if num > 255:
         num = 0
      self.current_level = num
      self.set_current_level_gfx ()
   def get_current_level_gfx (self):
      return self.gfx_current_level
   def set_current_level_gfx (self):
      self.gfx_current_level = convert_num_to_gfx (self.current_level + 1, 100, 3)

   def get_score (self):
      return self.score
   def set_score (self, num):
      self.score = num
      self.set_score_gfx ()
   def get_score_gfx (self):
      return self.gfx_score
   def set_score_gfx (self):
      prior = self.gfx_score[1]
      self.gfx_score = convert_num_to_gfx (self.score, 100000, 6)
      # new life every 10000 points
      if (prior != self.gfx_score[1]):
         # max is 255 lives
         if (get_lives (self) < 0xff):
            set_lives (self, get_lives (self) + 1)

   def get_time (self):
      return self.time
   def set_time (self, num):
      self.time = num
      self.set_time_gfx ()
   def get_time_gfx (self):
      return self.gfx_time
   def set_time_gfx (self):
      self.gfx_time = convert_num_to_gfx (self.time, 100, 3)

   def get_bonus (self):
      return (self.bonus)
   def set_bonus (self, num):
      self.bonus = num
      self.set_bonus_gfx ()
   def get_bonus_gfx (self):
      return self.gfx_bonus
   def set_bonus_gfx (self):
      self.gfx_bonus = convert_num_to_gfx (self.bonus, 1000, 4)

   def get_lives (self):
      return self.lives
   def set_lives (self, num):
      self.lives = num

   def get_sandbox (self, x, y):
      if (x,y) in self.sandbox:
         return self.sandbox[(x, y)]
      else:
         return 0
   def set_sandbox (self, x, y, content):
      self.sandbox[(x, y)] = content

   def get_n_eggs (self):
      return (self.n_of_eggs)
   def set_n_eggs (self, num):
      self.n_of_eggs = num

def convert_num_to_gfx (num, base, level):
   gfx = []
   for i in range (0, level):
      gfx.append (int (num / base))
      num -= gfx[i] * base
      base /= 10
   return gfx
